### PR TITLE
19 disk encryption support

### PR DESCRIPTION
**Pull request by alegrey91**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -451,12 +453,15 @@ Let's take a look at how the *inventory* file is going to be fill.
                             		pool: default                   # Store the disk image into the pool = default.
                             		format: xfs                     # Disk Filesystem = xfs
                             		mount_point: /mnt/data_storage  # The path where the disk is mounted
+                            		encryption: True                # Enable disk encryption
+                            		keyfile: ./passphrase.txt       # Specify the local file which contain the passphrase to encrypt the disk
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply README.md]: Done

File: roles/terraform/tasks/storage/disk_Debian_setup.yml [hunk 2]
```diff
@@ -19,6 +36,33 @@
   with_items:
     - "{{ ansible_devices.keys() | select('match','vd[b-z]') | list }}"
 
+# Copy the keyfile to the remote host, to allow 
+# luks utility to encrypt disks.
+- name: Copy keyfile to remote host
+  copy:
+    src: "{{ item.0['keyfile'] }}"
+    dest: "/tmp/{{ item.0['keyfile'] | basename }}"
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Debian_setup.yml]: moved under /root/keyfiles

File: roles/terraform/tasks/storage/disk_Debian_setup.yml [hunk 3]
```diff
@@ -19,6 +36,33 @@
...
+    - "{{ item.0['encryption'] }}"
+
+# For each partition create a luks container
+# using the specified keyfile to encrypt it.
+- name: Create LUKS container
+  luks_device:
+    name: "enc{{ item.1 }}1"
+    device: "/dev/{{ item.1 }}1"
+    state: "opened"
+    keyfile: "/tmp/{{ item.0['keyfile'] | basename }}"
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Debian_setup.yml]: as above

File: roles/terraform/tasks/storage/disk_Debian_setup.yml [hunk 4]
```diff
@@ -57,6 +117,36 @@
...
+    opts: nofail
+  with_together:
+    - "{{ data_disks.values() | list }}"
+    - "{{ ansible_devices.keys() | select('match','vd[b-z]') | sort | list }}"
+  when:
+    - "{{ item.0['encryption'] }}"
+
+# Remove the keyfile which contains the passphrase to 
+# access the disk.
+- name: Remove keyfiles
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Debian_setup.yml]: moved under /root/keyfiles and recorded in /etc/crypttab

File: roles/terraform/tasks/storage/disk_RedHat_setup.yml [hunk 5]
```diff
@@ -57,6 +117,36 @@
...
+    opts: nofail
+  with_together:
+    - "{{ data_disks.values() | list }}"
+    - "{{ ansible_devices.keys() | select('match','vd[b-z]') | sort | list }}"
+  when:
+    - "{{ item.0['encryption'] }}"
+
+# Remove the keyfile which contains the passphrase to 
+# access the disk.
+- name: Remove keyfiles
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_RedHat_setup.yml]: as above

File: roles/terraform/tasks/storage/disk_Suse_setup.yml [hunk 6]
```diff
@@ -19,6 +36,33 @@
   with_items:
     - "{{ ansible_devices.keys() | select('match','vd[b-z]') | list }}"
 
+# Copy the keyfile to the remote host, to allow 
+# luks utility to encrypt disks.
+- name: Copy keyfile to remote host
+  copy:
+    src: "{{ item.0['keyfile'] }}"
+    dest: "/tmp/{{ item.0['keyfile'] | basename }}"
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Suse_setup.yml]: as above

File: roles/terraform/tasks/storage/disk_Suse_setup.yml [hunk 7]
```diff
@@ -19,6 +36,33 @@
...
+    - "{{ item.0['encryption'] }}"
+
+# For each partition create a luks container
+# using the specified keyfile to encrypt it.
+- name: Create LUKS container
+  luks_device:
+    name: "enc{{ item.1 }}1"
+    device: "/dev/{{ item.1 }}1"
+    state: "opened"
+    keyfile: "/tmp/{{ item.0['keyfile'] | basename }}"
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Suse_setup.yml]: as above

File: roles/terraform/tasks/storage/disk_Suse_setup.yml [hunk 8]
```diff
@@ -57,6 +117,36 @@
...
+    opts: nofail
+  with_together:
+    - "{{ data_disks.values() | list }}"
+    - "{{ ansible_devices.keys() | select('match','vd[b-z]') | sort | list }}"
+  when:
+    - "{{ item.0['encryption'] }}"
+
+# Remove the keyfile which contains the passphrase to 
+# access the disk.
+- name: Remove keyfiles
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Suse_setup.yml]: as above

File: vars/os/family/Debian.yml [hunk 9]
```diff
@@ -2,6 +2,7 @@ ansible_dependencies:
   - python3
   - unattended-upgrades
   - parted
+  - cryptsetup
```
[the remark this answers is not in the document]
Comment (alegrey91) [reply vars/os/family/Debian.yml]: Your bad deads, I've done it.

File: roles/terraform/tasks/storage/disk_Debian_setup.yml [hunk 10]
```diff
@@ -19,6 +36,75 @@
...
+    - "{{ data_disks.values() | list }}"
+    - "{{ ansible_devices.keys() | select('match','vd[b-z]') | sort | list }}"
+    - "{{ data_disks.keys() | list }}"
+  when:
+    - "{{ item.0['encryption'] }}"
+
+# Generate keyfiles using /dev/urandom and store them 
+# into a local directory
+- name: Generate keyfiles
+  shell: |
```
Comment (89luca89) [inline roles/terraform/tasks/storage/disk_Debian_setup.yml:69]: shell must be idempotent, use exit codes and `changed_when` to ensure an `ok` when nothing is changed

File: roles/terraform/tasks/storage/disk_RedHat_setup.yml [hunk 11]
```diff
@@ -19,6 +36,75 @@
...
+    - "{{ data_disks.values() | list }}"
+    - "{{ ansible_devices.keys() | select('match','vd[b-z]') | sort | list }}"
+    - "{{ data_disks.keys() | list }}"
+  when:
+    - "{{ item.0['encryption'] }}"
+
+# Generate keyfiles using /dev/urandom and store them 
+# into a local directory
+- name: Generate keyfiles
+  shell: |
```
Comment (89luca89) [inline roles/terraform/tasks/storage/disk_RedHat_setup.yml:69]: shell must be idempotent, use exit codes and changed_when to ensure an ok when nothing is changed

File: roles/terraform/tasks/storage/disk_Suse_setup.yml [hunk 12]
```diff
@@ -19,6 +36,75 @@
...
+    - "{{ data_disks.values() | list }}"
+    - "{{ ansible_devices.keys() | select('match','vd[b-z]') | sort | list }}"
+    - "{{ data_disks.keys() | list }}"
+  when:
+    - "{{ item.0['encryption'] }}"
+
+# Generate keyfiles using /dev/urandom and store them 
+# into a local directory
+- name: Generate keyfiles
+  shell: |
```
Comment (89luca89) [inline roles/terraform/tasks/storage/disk_Suse_setup.yml:69]: shell must be idempotent, use exit codes and changed_when to ensure an ok when nothing is changed

**Review by 89luca89 — CHANGES_REQUESTED** [review]

**Comment by alegrey91** [comment]
I've found a bug during some tests.
This happens when a VM (or more) in the inventory is declared without the `data_disks` keyword.
The error message is: AnsibleUndefinedVariable: 'data_disks' is undefined.
I'll push the fix asap.

File: roles/terraform/tasks/storage/disk_Debian_setup.yml (hunk 10 above)
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Debian_setup.yml:69]: solved

File: roles/terraform/tasks/storage/disk_RedHat_setup.yml (hunk 11 above)
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_RedHat_setup.yml:69]: solved as above

File: roles/terraform/tasks/storage/disk_Suse_setup.yml (hunk 12 above)
Comment (alegrey91) [reply roles/terraform/tasks/storage/disk_Suse_setup.yml:69]: solved as above

**Pull request merged**